### PR TITLE
test-order

### PR DESCRIPTION
**Pull request by khungking909**
## Related Tickets
- [#74351](https://edu-redmine.sun-asterisk.vn/issues/74351)

## WHAT (optional)
- Test admin orders controller

## Evidence (Screenshot or Video)
image

image

image

### CODE REVIEW

**Comment by khungking909** [comment]
ready

File: spec/requests/admin/orders_controller_spec.rb [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    let!(:orders) { create_list(:order, 4, account_id: user.id) }
+
+    context "when login success and render success" do
+      include_context "with a logged-in with admin", :get, :index
+
+      it "render index" do
+        expect(response).to(render_template("index"))
+        expect(assigns(:orders)).to(all(be_a(Order)))
+        expect(assigns(:orders).count).to(eq(2))
+        expect(assigns(:pagy).count).to(eq(Order.count))
```
Comment (binhpt-3177) [inline spec/requests/admin/orders_controller_spec.rb]:
```suggestion
        expect(assigns(:pagy).count).to(eq(4))
```

File: spec/requests/admin/orders_controller_spec.rb [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    let!(:order_histories) { create_list(:order_history, 6, order_id: order.id, product_id: product.id) }
+
+    context "when login success" do
+      include_context "with a logged-in with admin", :get, :show
+
+      context "render show sucess" do
+        it "render show success" do
+          expect(response).to(render_template("show"))
+          expect(assigns(:order_products)).to(all(be_a(OrderHistory)))
+          expect(assigns(:order_products).count).to(eq(5))
```
Comment (binhpt-3177) [inline spec/requests/admin/orders_controller_spec.rb:60]: Check total price có trả về đúng không

File: spec/requests/admin/orders_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    include_context "login fails", :get, :show
+  end
+
+  describe "PATCH #update" do
+    context "when login success and render success" do
+      include_context "with a logged-in with admin", :patch, :update
+
+      context "when order.status = 'pending' is " do
+        it "sets flash[:admin_success] with 'accept_order' message" do
+          @order_pending_accept = create(:order, account_id: account.id)
```
Comment (binhpt-3177) [inline spec/requests/admin/orders_controller_spec.rb]: Đặt ra let

File: spec/requests/admin/orders_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+          patch :update, params: { id: @order_pending_accept.id, status: "accept", comment: "accept" }
```
Comment (binhpt-3177) [inline spec/requests/admin/orders_controller_spec.rb]:
```suggestion
          patch :update, params: { id: @order_pending_accept.id, status: :accept, comment: :accept }
```

File: spec/requests/admin/orders_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+          expect(assigns(:order).status_approved?).to(eq(true))
```
Comment (binhpt-3177) [inline spec/requests/admin/orders_controller_spec.rb]:
```suggestion
          expect(assigns(:order).reload.status_approved?).to(eq(true))
```

File: spec/requests/admin/orders_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+          expect(flash[:admin_success]).to(eq(I18n.t("orders.accept_order", name: @order_pending_accept.receiver_name)))
+          expect(response).to(redirect_to(admin_orders_path))
+        end
+
+        it "sets flash[:admin_success] with 'refuse_order' message" do
+          @order_pending_refuse = create(:order, account_id: account.id)
+          patch :update, params: { id: @order_pending_refuse.id, status: "other_status_difficult_accept", comment: "refuse" }
```
Comment (binhpt-3177) [inline spec/requests/admin/orders_controller_spec.rb]: Dùng symbor

**Comment by khungking909** [comment]
ready

File: spec/requests/admin/orders_controller_spec.rb [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+          expect(assigns(:order_products).count).to(eq(5))
+          expect(total_price(assigns(:order))).to(eq(3000))
+          expect(assigns(:pagy).count).to(eq(6))
+          expect(response).to(have_http_status(:success))
+        end
+      end
+
+      context "render show fails" do
+        it "render show fails because order not found" do
+          get :show, params: { id: 9_999_999_999_999_999 }
```
Comment (binhpt-3177) [inline spec/requests/admin/orders_controller_spec.rb]:
```suggestion
          get :show, params: { id: -1 }
```

File: spec/requests/admin/orders_controller_spec.rb [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+
+          expect(assigns(:order).reload.status).to(eq(order_pending_approved.status))
+          expect(flash[:admin_error]).to(eq(I18n.t("orders.status_fails")))
+          expect(response).to(redirect_to(admin_orders_path))
+        end
+      end
+
+      context "when update fails" do
+        it "when update fails because was not found order" do
+          patch :update, params: { id: 9_999_999_999, status: :reject }
```
Comment (binhpt-3177) [inline spec/requests/admin/orders_controller_spec.rb]: .

File: spec/requests/admin/products_controller_spec.rb [hunk 9]
```diff
@@ -209,7 +209,19 @@ def perform_action(method, action)
...
           expect(flash[:admin_error]).to(eq(I18n.t("admin.products.load.not_found")))
           expect(response).to(have_http_status(:found))
-          expect(response).to(redirect_to(admin_orders_path))
+          expect(response).to(redirect_to(admin_products_path))
+        end
+
+        it "delete fails with product of order status pending" do
+          @product = create(:product, category_id: category.id)
+          @order = create(:order, account_id: current_account.id)
+          @order_history = create(:order_history, order_id: @order.id, product_id: @product.id)
```
Comment (binhpt-3177) [inline spec/requests/admin/products_controller_spec.rb]: .

**Comment by khungking909** [comment]
ready

**Review by binhpt-3177 — APPROVED** [review]

**Pull request merged**